hive/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveTableUtils.java

```diff
@@ -50,33 +56,48 @@ static Schema getSchema(org.apache.hadoop.hive.metastore.api.Table table) {
     Schema schema;
     if (schemaStr == null) {
       LOG.warn("Table {}.{} does not have an avro.schema.literal set; using Hive schema instead. The schema will not" +
-          " have case sensitivity and nullability information", table.getDbName(), table.getTableName());
+                   " have case sensitivity and nullability information", table.getDbName(), table.getTableName());
       // TODO: Add support for tables without avro.schema.literal
       throw new UnsupportedOperationException("Reading tables without avro.schema.literal not implemented yet");
     } else {
       schema = AvroSchemaUtil.toIceberg(new org.apache.avro.Schema.Parser().parse(schemaStr));
     }
 
-    List<String> partCols = table.getPartitionKeys().stream().map(FieldSchema::getName).collect(Collectors.toList());
-    return addPartitionColumnsIfRequired(schema, partCols);
-  }
+    Types.StructType dataStructType = schema.asStruct();
+    List<Types.NestedField> fields = Lists.newArrayList(dataStructType.fields());
 
-  private static Schema addPartitionColumnsIfRequired(Schema schema, List<String> partitionColumns) {
-    List<Types.NestedField> fields = new ArrayList<>(schema.columns());
-    AtomicInteger fieldId = new AtomicInteger(10000);
-    partitionColumns.stream().forEachOrdered(column -> {
-      Types.NestedField field = schema.findField(column);
-      if (field == null) {
```

Comment:
There are tables at LI which have datepartition in both data columns and partition columns, hence I had this check to only add the columns to the schema if they don't already have the partition column in the schema. Can we keep this check in the new logic?

Reply:
Should we throw an exception instead? There could be some weird cases like
* What if the fields with the same name are of different types?
* What if the type is the same but have different values? 2020323 vs 2020-03-23 or something with completely different semantics?

Reply:
I did a quick query and seems like only 25 user tables and 1 metrics table are affected. So I am okay throwing an exception.

Reply:
Thanks for the analysis.